distributed/worker.py

```diff
@@ -1452,19 +1462,30 @@ def add_task(
                         "Asked to compute pre-existing result: %s: %s", key, ts.state
                     )
                     self.send_task_state_to_scheduler(ts)
+                    # FIXME: Can we pull this off without a return to ensure
+                    # that we do not miss anything following?
                     return
                 if ts.state in IN_PLAY:
-                    return
+                    # XXX I believe we only require this special treatment in
+                    # case we're already executing or in memory to avoid
+                    # resetting / aborting the computation. Then again, if the
+                    # execution is ongoing, finishing it _should_ transfer it to
+                    # memory regardless of the start state, shouldn't it?
+
+                    # State waiting is ambiguous since it is used for
+                    # dependencies which are waiting to be fetched (no runspec)
+                    # and for tasks ready to be computed (with runspec)
+                    if ts.state == "waiting":
+                        ts.runspec = runspec
                 if ts.state == "erred":
                     ts.exception = None
                     ts.traceback = None
                 else:
-                    ts.state = "waiting"
+                    self.transition(ts, "waiting", runspec=runspec)
```

Comment:
`start == finish` is treated in Worker.transition

https://github.com/dask/distributed/blob/3fb3d8c8c14800ff8f1cd76d0052985f355ecfcc/distributed/worker.py#L1532-L1537

This might still not be correct, though, since `waiting` is currently ambiguous since it refers to a dependency waiting to be computed and fetched and simultaneously to a runnable task. Both cases allow for very different transitions.

This ambiguity doesn't feel correct and I'm wondering if there is something we can do about this. IMHO, the `state` attribute should uniquely define which transition a task is allowed to perform and which not but this is currently not the case. I was already thinking about introducing another state, like `waiting_runnable` to deal with this ambiguity but I would like to only introduce a new state if absolutely necessary

Reply:
> `start == finish` is treated in Worker.transition

Ahh, yeah.  

> currently ambiguous since it refers to a dependency waiting to be computed and fetched and simultaneously to a runnable task.

yeah, this is ambiguous.  Ideally tasks that are runnable should be marked `ready`, but the default starting state is `waiting`, which leads to this ambiguity.  We might be better served by having a default task state of `new` (as you referenced in a code comment, there are no `new->?` transition functions but that might be worth doing.
I can try to put together a draft of what that might look like and see how it works (unless you're interested in tackling that yourself)

Reply:
>  Ideally tasks that are runnable should be marked ready

I believe it's a bit more tricky than this. I'm not necessarily talking about "ready" tasks in a sense that they can be calculated *now* but with "runnable" I am referring to tasks, this worker is supposed to calculate at some point vs dependencies the worker is only supposed to fetch. This might include speculative assignments (let's include this for a moment since this is still WIP) but also tasks which were ready at some point but one of their dependencies got lost while the worker was trying to fetch them. Therefore, there must be a state for _runnable_ tasks before the `ready` state. On scheduler side this is `released` (although imo `released` is not the best name for this state)

I believe it is worth the effort putting more though into the state mapping and a new state might be the way to go here (not sure, whether it should be called "new" or not)

I am trying to stabilize this PR without introducing any new state(s). I found another race condition in the way missing deps are handled and I believe this is doable without a new state. I would still propose to discuss this additional state topic but on a different PR/issue. If you already have an idea, feel free to suggest something. We might also want to look into more documentation of what the states are _supposed_ to do (also out of scope for this PR)